pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompactmerge/UpsertCompactMergeTaskGenerator.java

```diff
@@ -0,0 +1,424 @@
+import org.apache.pinot.common.utils.SegmentUtils;
+import org.apache.pinot.controller.helix.core.PinotHelixResourceManager;
+import org.apache.pinot.controller.helix.core.minion.generator.BaseTaskGenerator;
+import org.apache.pinot.controller.helix.core.minion.generator.TaskGeneratorUtils;
+import org.apache.pinot.controller.util.ServerSegmentMetadataReader;
+import org.apache.pinot.core.common.MinionConstants;
+import org.apache.pinot.core.minion.PinotTaskConfig;
+import org.apache.pinot.spi.annotations.minion.TaskGenerator;
+import org.apache.pinot.spi.config.table.TableConfig;
+import org.apache.pinot.spi.config.table.TableType;
+import org.apache.pinot.spi.config.table.UpsertConfig;
+import org.apache.pinot.spi.utils.TimeUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+@TaskGenerator
+public class UpsertCompactMergeTaskGenerator extends BaseTaskGenerator {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(UpsertCompactMergeTaskGenerator.class);
+  private static final String DEFAULT_BUFFER_PERIOD = "2d";
+  private static final int DEFAULT_NUM_SEGMENTS_BATCH_PER_SERVER_REQUEST = 500;
+
+  public static class SegmentMergerMetadata {
+    private final SegmentZKMetadata _segmentZKMetadata;
+    private final long _validDocIds;
+    private final long _invalidDocIds;
+
+    SegmentMergerMetadata(SegmentZKMetadata segmentZKMetadata, long validDocIds, long invalidDocIds) {
+      _segmentZKMetadata = segmentZKMetadata;
+      _validDocIds = validDocIds;
+      _invalidDocIds = invalidDocIds;
+    }
+
+    public SegmentZKMetadata getSegmentZKMetadata() {
+      return _segmentZKMetadata;
+    }
+
+    public long getValidDocIds() {
+      return _validDocIds;
+    }
+
+    public long getInvalidDocIds() {
+      return _invalidDocIds;
+    }
+  }
+
+  public static class SegmentSelectionResult {
+
+    private final Map<Integer, List<List<SegmentMergerMetadata>>> _segmentsForCompactMergeByPartition;
+
+    private final List<String> _segmentsForDeletion;
+
+    SegmentSelectionResult(Map<Integer, List<List<SegmentMergerMetadata>>> segmentsForCompactMergeByPartition,
+        List<String> segmentsForDeletion) {
+      _segmentsForCompactMergeByPartition = segmentsForCompactMergeByPartition;
+      _segmentsForDeletion = segmentsForDeletion;
+    }
+
+    public Map<Integer, List<List<SegmentMergerMetadata>>> getSegmentsForCompactMergeByPartition() {
+      return _segmentsForCompactMergeByPartition;
+    }
+
+    public List<String> getSegmentsForDeletion() {
+      return _segmentsForDeletion;
+    }
+  }
+
+  @Override
+  public String getTaskType() {
+    return MinionConstants.UpsertCompactMergeTask.TASK_TYPE;
+  }
+
+  @Override
+  public List<PinotTaskConfig> generateTasks(List<TableConfig> tableConfigs) {
+    String taskType = MinionConstants.UpsertCompactMergeTask.TASK_TYPE;
+    List<PinotTaskConfig> pinotTaskConfigs = new ArrayList<>();
+    for (TableConfig tableConfig : tableConfigs) {
+
```

Comment:
nit: move the for-loop logic to a helper method to be more readable

Reply:
yes agreed! let me do this in a follow-up.